pkg/skuba/actions/auth/auth_test.go

```diff
@@ -0,0 +1,521 @@
+				"kid": jwk.KeyID,
+				"n":   n(&key.PublicKey),
+				"e":   e(&key.PublicKey),
+			}},
+		})
+	}
+}
+
+func ApprovalHandler() func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		htmlOutput := fmt.Sprintf(`
+			<div class="theme-panel">
+			<h2 class="theme-heading">Login Successful</h2>
+			<p>Please copy this code, switch to your application and paste it there:</p>
+			<input type="text" class="theme-form-input" value="%s" />
+			</div>
+		`, newID())
+		_, _ = w.Write([]byte(htmlOutput))
+	}
+}
+
+func ApprovalInvalidBodyHandler() func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		htmlOutput := fmt.Sprintf(`
+			<div class="theme-panel">
+			<h2 class="theme-heading">Login Successful</h2>
+			</div>
+		`)
+		_, _ = w.Write([]byte(htmlOutput))
+	}
+}
+
+func Test_doAuth(t *testing.T) {
+	tests := []struct {
+		name               string
+		srvCb              func() *httptest.Server
+		clientID           string
+		clientSecret       string
+		issuerURL          string
+		username           string
+		password           string
+		rootCAData         []byte
+		insecureSkipVerify bool
+		expectedError      bool
+	}{
+		{
+			name: "login with root CA provided",
+			srvCb: func() *httptest.Server {
+				mux := http.NewServeMux()
+				mux.HandleFunc("/.well-known/openid-configuration", OpenIDHandler())
+				mux.HandleFunc("/auth", AuthHandler())
+				mux.HandleFunc("/auth/local", AuthLocalHandler())
+				mux.HandleFunc("/token", TokenHandler())
+				mux.HandleFunc("/keys", KeysHandler())
+				mux.HandleFunc("/approval", ApprovalHandler())
+
+				srv := httptest.NewUnstartedServer(mux)
+				cert, _ := tls.X509KeyPair(LocalhostCert, LocalhostKey)
+				srv.TLS = &tls.Config{
+					Certificates: []tls.Certificate{cert},
+				}
+				srv.StartTLS()
+				return srv
+			},
+			clientID:     clientID,
+			clientSecret: clientSecret,
+			username:     MockDefaultUsername,
+			password:     MockDefaultPassword,
+			rootCAData:   LocalhostCert,
+		},
+		{
+			name: "login with no root CA provided",
+			srvCb: func() *httptest.Server {
+				mux := http.NewServeMux()
+				mux.HandleFunc("/.well-known/openid-configuration", OpenIDHandler())
+				mux.HandleFunc("/auth", AuthHandler())
+				mux.HandleFunc("/auth/local", AuthLocalHandler())
+				mux.HandleFunc("/token", TokenHandler())
+				mux.HandleFunc("/keys", KeysHandler())
+				mux.HandleFunc("/approval", ApprovalHandler())
+
+				srv := httptest.NewUnstartedServer(mux)
+				cert, _ := tls.X509KeyPair(LocalhostCert, LocalhostKey)
+				srv.TLS = &tls.Config{
+					Certificates: []tls.Certificate{cert},
+				}
+				srv.StartTLS()
+				return srv
+			},
+			clientID:      clientID,
+			clientSecret:  clientSecret,
+			username:      MockDefaultUsername,
+			password:      MockDefaultPassword,
+			rootCAData:    invalidCert,
+			expectedError: true,
+		},
+		{
+			name: "login with insecure skip verify",
```

Comment:
Covered in Test_Login.

Reply:
Remove `Test_doAuth` to avoids duplicate testing.